### PR TITLE
feat: disk usage monitoring for readonly mode

### PR DESCRIPTION
**Pull request by pcnc**
## What kind of change does this PR introduce?

* introduces `monitor` command
* currently monitors only data disk usage, enabling readonly mode

Required by [postgres#312](https://github.com/supabase/postgres/pull/312)

### CODE REVIEW

File: monitor/diskUsage.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	var fsStats unix.Statfs_t
+
+	err := unix.Statfs(d.dataDiskPath, &fsStats)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get filesystem stats.")
+	}
+
+	usedBlocksPct := (1 - float64(fsStats.Bavail)/float64(fsStats.Blocks))
+
+	if usedBlocksPct >= 0.97 {
```
Comment (darora) [inline monitor/diskUsage.go:66]: configurable?

File: monitor/diskUsage.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	var dataDiskPath string
+	_, err := os.Stat("/data")
+	if os.IsNotExist(err) {
+		dataDiskPath = "/"
+	} else {
+		dataDiskPath = "/data"
+	}
+
+	return &DiskUsageMonitor{
+		interval:            5 * time.Second,
```
Comment (darora) [inline monitor/diskUsage.go:31]: also configurable

**Review by darora — APPROVED** [review]
list of monitors being configurable would also save us a bunch of hassle if one of them turned out to be troublesome

**Comment by pcnc** [comment]
> list of monitors being configurable would also save us a bunch of hassle if one of them turned out to be troublesome

And done

File: monitors/diskUsage.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+func (d *DiskUsageMonitor) readOnlyModeOverrideExists() (bool, error) {
+	cmd := exec.Command("/bin/sh", "-c", "sudo /root/manage_readonly_mode.sh check_override")
+
+	output, err := cmd.Output()
+	if err != nil {
+		return false, errors.Wrapf(err, "Couldn't check readonly mode override presence: %s", output)
+	}
+
+	if strings.TrimSuffix(string(output), "\n") == "1" {
```
Comment (pcnc) [inline monitors/diskUsage.go:130]: Couldn't get script to only output value, sans new line

Comment (darora) [reply monitors/diskUsage.go:130]: Not that I really care about changing this, but an `echo -n $RESULT` should do the trick?

**Review by darora — APPROVED** [review]

File: monitors/diskUsage.go (hunk 3 above)
Comment (pcnc) [reply monitors/diskUsage.go:130]: That's.. very sane and fair - thank you

**Pull request merged**